File: src/laohuangcode/cancellation.py
"""Cooperative, task-scoped cancellation primitives."""

from __future__ import annotations

from collections.abc import Callable
from threading import Event, Lock
from time import monotonic


class CancellationError(RuntimeError):
    """Raised by :meth:`CancelToken.throw_if_cancelled`."""


CancelCallback = Callable[[str], None]


class CancelToken:
    """A thread-safe, idempotent cancellation token.

    Registered callbacks run synchronously in the thread that wins the first
    call to :meth:`cancel`. A callback registered after cancellation runs
    immediately in the registering thread.
    """

    def __init__(self) -> None:
        self._event = Event()
        self._lock = Lock()
        self._reason: str | None = None
        self._requested_at: float | None = None
        self._callbacks: dict[int, CancelCallback] = {}
        self._callback_id = 0

    @property
    def reason(self) -> str | None:
        with self._lock:
            return self._reason

    @property
    def requested_at(self) -> float | None:
        with self._lock:
            return self._requested_at

    @property
    def cancelled(self) -> bool:
        return self._event.is_set()

    def is_cancelled(self) -> bool:
        return self._event.is_set()

    def cancel(self, reason: str = "cancelled") -> bool:
        if not reason:
            reason = "cancelled"
        with self._lock:
            if self._event.is_set():
                return False
            self._reason = reason
            self._requested_at = monotonic()
            callbacks = tuple(self._callbacks.values())
            self._callbacks.clear()
            self._event.set()
        for callback in callbacks:
            try:
                callback(reason)
            except Exception:
                # Cancellation must remain effective even if a cleanup hook is
                # faulty; owners can report their own cleanup failures.
                continue
        return True

    def wait(self, timeout: float | None = None) -> bool:
        return self._event.wait(timeout)

    def throw_if_cancelled(self) -> None:
        if self._event.is_set():
            raise CancellationError(self.reason or "cancelled")

    def register(self, callback: CancelCallback) -> Callable[[], None]:
        """Register a cancellation hook and return an unregister function."""

        with self._lock:
            if self._event.is_set():
                reason = self._reason or "cancelled"
                callback_id = None
            else:
                self._callback_id += 1
                callback_id = self._callback_id
                self._callbacks[callback_id] = callback
                reason = None
        if reason is not None:
            try:
                callback(reason)
            except Exception:
                pass

        def unregister() -> None:
            if callback_id is None:
                return
            with self._lock:
                self._callbacks.pop(callback_id, None)

        return unregister

File: packages/core/runtime-protocol/src/cancellation.ts
/** Cooperative, task-scoped cancellation primitives. */

/** Raised by {@link CancelToken.throwIfCancelled}. */
export class CancellationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "CancellationError";
  }
}

export type CancelCallback = (reason: string) => void;

/**
 * An idempotent, single-threaded cancellation token.
 *
 * Registered callbacks run synchronously in the call stack that wins the
 * first call to {@link cancel}. A callback registered after cancellation
 * runs immediately, before {@link register} returns.
 */
export class CancelToken {
  #cancelled = false;
  #reason: string | null = null;
  #requestedAt: number | null = null;
  #callbacks = new Map<number, CancelCallback>();
  #nextCallbackId = 0;
  #waiters = new Set<() => void>();
  #controller = new AbortController();

  get reason(): string | null {
    return this.#reason;
  }

  /** `performance.now()` timestamp of the first cancel, or null. */
  get requestedAt(): number | null {
    return this.#requestedAt;
  }

  get cancelled(): boolean {
    return this.#cancelled;
  }

  /** AbortSignal view of the token, for APIs that accept one. */
  get signal(): AbortSignal {
    return this.#controller.signal;
  }

  isCancelled(): boolean {
    return this.#cancelled;
  }

  /**
   * Request cancellation. Returns true for the first (winning) call and
   * false for every subsequent call.
   */
  cancel(reason = "cancelled"): boolean {
    const effectiveReason = reason || "cancelled";
    if (this.#cancelled) {
      return false;
    }
    this.#reason = effectiveReason;
    this.#requestedAt = performance.now();
    const callbacks = [...this.#callbacks.values()];
    this.#callbacks.clear();
    this.#cancelled = true;
    this.#controller.abort(new CancellationError(effectiveReason));
    for (const callback of callbacks) {
      try {
        callback(effectiveReason);
      } catch {
        // Cancellation must remain effective even if a cleanup hook is
        // faulty; owners can report their own cleanup failures.
        continue;
      }
    }
    for (const resolve of this.#waiters) {
      resolve();
    }
    this.#waiters.clear();
    return true;
  }

  /**
   * Resolve true once the token is cancelled. With `timeoutMs`, resolve
   * false instead when the timeout elapses first.
   */
  wait(timeoutMs?: number): Promise<boolean> {
    if (this.#cancelled) {
      return Promise.resolve(true);
    }
    return new Promise<boolean>((resolve) => {
      let timer: NodeJS.Timeout | undefined;
      const onCancel = (): void => {
        if (timer !== undefined) {
          clearTimeout(timer);
        }
        this.#waiters.delete(onCancel);
        resolve(true);
      };
      this.#waiters.add(onCancel);
      if (timeoutMs !== undefined) {
        timer = setTimeout(() => {
          this.#waiters.delete(onCancel);
          resolve(false);
        }, timeoutMs);
      }
    });
  }

  throwIfCancelled(): void {
    if (this.#cancelled) {
      throw new CancellationError(this.#reason ?? "cancelled");
    }
  }

  /** Register a cancellation hook and return an unregister function. */
  register(callback: CancelCallback): () => void {
    if (this.#cancelled) {
      const reason = this.#reason ?? "cancelled";
      try {
        callback(reason);
      } catch {
        // Late-registered hooks must not break the registering caller.
      }
      return () => {};
    }
    this.#nextCallbackId += 1;
    const callbackId = this.#nextCallbackId;
    this.#callbacks.set(callbackId, callback);
    return () => {
      this.#callbacks.delete(callbackId);
    };
  }
}
